### ISSUE TITLE
Proposal for join_if

### ISSUE THREAD

**Issue by RP-pl** [more-itertools/more-itertools#833]
### Description
The proposed itertool joins a new iterable with the previously joined iterables when it meets a specific condition. This condition takes two arguments: the first argument is an iterable made out of the previously joined iterables, and the second argument is the currently tested iterable.

### References
Had data provided in batches, but not all of the batches had features that would be useful, so I needed to pick only appropriate data

### Examples
```python
def lower_variance(batches):
    """Pick only the batches, where the variance is lower than the overall variance."""
    return join_if(batches, predicate=lambda x,y: np.var(x) > np.var(y))

def join_with_filter(iterables, filter_func=lambda x: len(x) == 2):
    """Shorthand for filter and join. Join only the iterables that pass the filter."""
    return join_if(iterables, predicate=lambda x,y: filter_func(y), test_first=True)
```

**Comment by JamesParrott** [comment]
Python already has 'filter' and 'zip'.  Please can you give an example of what you mean by "joining" an iterable?

**Comment by RP-pl** [comment]
By joining an iterables i mean creating a single collection out of them.
For example:
```python
join_if([1,1,1], [2,2,2],[4,4,4],[5,5,5], [3,3,3], predicate=lambda x,y: np.average(y) <= 3, test_first=True)
```
should return 
```
[1,1,1,2,2,2,3,3,3]
```

As for an example with sequential joining:

```python
join_if([1,1,1], [2,2,2],[4,4,4],[5,5,5], [3,3,3], predicate=lambda x,y: x[-1] < y[-1])
```
should return
```
[1, 1, 1, 2, 2, 2, 4, 4, 4, 5, 5, 5]
```

**Comment by JamesParrott** [comment]
I'm not sure how the predicate taking two elements, testing the last element, gives both the 1s and the 3s.  
Nor why the 5s are there in your second example.

Something like this, that avoids those issues can be achieved with itertools.chain, more_itertools.first, filter, and itertools.pairwise.  

The name `chain_if` would be more consistent with the existing tools.

**Comment by pochmann3** [comment]
@JamesParrott I think they mean this:

```python
def join_if(*iterables, predicate, test_first=False):
    joined = []
    for it in iterables:
        current = list(it)
        if not test_first and not joined or predicate(joined, current):
            joined += current
    return joined

print(join_if([1,1,1], [2,2,2],[4,4,4],[5,5,5], [3,3,3], predicate=lambda x,y: x[-1] < y[-1]))
```

[Attempt This Online!](https://ato.pxeger.com/run?1=TVC7bsMwDERXfwVHq2WG9AEUQb127ywIgRJRKAtXNiQaiL-lS5b2o_o1lZ81byDEI4_Uff20vbw34Xr97sTvnn9v3hx5-Gg4HNmXtywU7ammhNBGcny2QghCSY6eY5Lq1daJ1KGAHMMUOahAm_HtmwgswAFWmalxiHMXIwXJ3TUnKVnUSrGH0MhmC9jgxtK8Ieuu15RTDRdB9b9ic9NdtfAjG0m6GGayKNrIQcrl03qPGQZB32OGQf2IGTk_YcZAPGCG2XhS1fbz5CxcsD_ARe_2Bl6gH7JSk7OzwYvRfw)

**Comment by RP-pl** [comment]
> @JamesParrott I think they mean this:
> 
> ```python
> def join_if(*iterables, predicate, test_first=False):
>     joined = []
>     for it in iterables:
>         current = list(it)
>         if not test_first and not joined or predicate(joined, current):
>             joined += current
>     return joined
> 
> print(join_if([1,1,1], [2,2,2],[4,4,4],[5,5,5], [3,3,3], predicate=lambda x,y: x[-1] < y[-1]))
> ```
> 
> [Attempt This Online!](https://ato.pxeger.com/run?1=TVC7bsMwDERXfwVHq2WG9AEUQb127ywIgRJRKAtXNiQaiL-lS5b2o_o1lZ81byDEI4_Uff20vbw34Xr97sTvnn9v3hx5-Gg4HNmXtywU7ammhNBGcny2QghCSY6eY5Lq1daJ1KGAHMMUOahAm_HtmwgswAFWmalxiHMXIwXJ3TUnKVnUSrGH0MhmC9jgxtK8Ieuu15RTDRdB9b9ic9NdtfAjG0m6GGayKNrIQcrl03qPGQZB32OGQf2IGTk_YcZAPGCG2XhS1fbz5CxcsD_ARe_2Bl6gH7JSk7OzwYvRfw)

That is mostly my contribution idea

**Comment by JamesParrott** [comment]
Thanks.  I didn't realise the predicate tests the entire chain up to the current point, along with the next iterable, not pairwise iterables.  If it's not the intention to call the predicate on the entire list to date, is it possible to write it as a .  I get it though, and it's a lot nicer than my attempt below:

```
import itertools as it
import more_itertools as mit
l=[[1,1,1], [2,2,2],[4,4,4],[5,5,5], [3,3,3]]
f = lambda pair: pair[0][-1] < pair[1][-1]
g = lambda pair: sum(pair[1]) <= len(pair[1])*3
print(list(it.chain(*(mit.first(pair) for pair in filter(f, it.pairwise(l))))))
print(list(it.chain(*(mit.first(pair) for pair in filter(g, it.pairwise(l))))))
```

**Comment by JamesParrott** [comment]
@RP-pl Yeah, I wondered if you could use functools.reduce, with a function that appends.  It's the same pattern, it's just not  -reducing- the size of the list.

**Comment by RP-pl** [comment]
@JamesParrott I was thinking about that and that would certainly work, but i think, that in-function state modification with condition checking would ruin the functional approach that map-reduce provides.

**Comment by pochmann3** [comment]
Some thoughts...

Wouldn't `def join_if(iterables, ...)` be better than `def join_if(*iterables, ...)`? I think it's more natural to have an iterable of iterables. Your two examples for example, what's up with all that packing/unpacking?:
```
def lower_variance(*batches):
    """Pick only the batches, where the variance is lower than the overall variance."""
    return join_if(*batches, predicate=lambda x,y: np.var(x) > np.var(y))

def join_with_filter(*iterables, filter_func=lambda x: len(x) == 2):
    """Shorthand for filter and join. Join only the iterables that pass the filter."""
    return join_if(*iterables, predicate=lambda x,y: filter_func(y), test_first=True)
```

I'm worried this tool will encourage inefficient algorithms, if the predicate actually always looks at the whole already joined part and something incremental is possible (not sure about `np.var`, but for example `sum` clearly doesn't have to always sum from scratch but could just keep a running sum).

What if `test_first=False` and the first irritable is empty? Should the second iterable then still be tested, even though `joined` is still empty?

**Comment by RP-pl** [comment]
Chainging `def join_if(*iterables, ...)` to `def join_if(iterables, ...)` is a good idea. 
Fixing examples is also great idea.

When it comes to inefficient algorithms in my opinion there is no simple answer to that.

With your example with `sum` programmer could always write code like that:
```python

partial_sum = 0

def check_if_sum(x,y):
    if sum(y) > sum(x):
        partial_sum += sum(y)
        return True
    return False

join_if(data,predicate=check_if_sum)

```

In my opinion it's up to programmer to decide how efficient code they should write.

In my opinion when `test_first=False` and first iterable is empty second iterable should still be tested.

**Comment by bbayles** [comment]
Thanks for the suggestion. Could you provide examples from StackOverflow or Github of people trying to solve this problem?

**Comment by RP-pl** [comment]
This function might be a lightweight solution for problems such as:

https://stackoverflow.com/questions/71024707/cleaning-up-data-with-variably-noisy-signal

Without using scipy we could do something like:
```python
def noise_filter(iterables, tol_scale, tol_offset):
    """Filter out noise
        tol_scale: tolerance scaling variance of (x)
        tol_offset: flat tolerance
    """
    return chain_if(iterables, predicate=lambda x,y: np.var(y) < tol_scale*np.var(x) + tol_offset)

filtered = noise_filter(data, tol_scale=0.01, tol_offset=0.01)
```

For such a simple code, this solution would provide quite an effective filtering, as shown on the plot below:

![plot_good](https://github.com/more-itertools/more-itertools/assets/68811106/ea275801-c166-4985-b80c-879cb223fbaa)

**Comment by bbayles** [comment]
Thanks for this info the the contribution. I'm undecided about including it in the library, and will continue to meditate upon it / keep a lookout for applications.

**Comment by rhettinger** [comment]
FWIW, I haven't found any applications for this since it was proposed last year.

Also the suggested filtering application looks dubious to me. The median-filter suggested in the StackOverlow answer seems like a more plausible way to go (because occasionally dropping data defeats Fourier analysis or anything else that matches data with specific points in time).
